### PR TITLE
[9.0][PORT] base_delivery_carrier_label

### PR DESCRIPTION
**Pull request by yvaucher**

### CODE REVIEW

**Comment by yvaucher** [comment]
@bealdav A little review ?

**Comment by bealdav** [comment]
It seems than improvements in v8 are not there like unit tests ?

**Comment by yvaucher** [comment]
@bealdav Ok this PR is quite old I'll check if something has been merged in the mean time

**Comment by yvaucher** [comment]
@bealdav Rebased on updated modules with last changes.

File: base_delivery_carrier_label/__openerp__.py [hunk 1]
```diff
@@ -1,40 +1,22 @@
...
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
-
+# © 2013-2015 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 {'name': 'Base module for carrier labels',
- 'version': '8.0.1.2.0',
+ 'version': '9.0.1.2.0',
```
Comment (guewen) [inline base_delivery_carrier_label/__openerp__.py]: `s/9.0.1.2.0/9.0.1.0.0/`

File: base_delivery_carrier_label/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+# © 2013-2015 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (guewen) [inline base_delivery_carrier_label/models/__init__.py]: not sure there is a point to leave those lines in `__init__.py` files

File: base_delivery_carrier_label/models/stock.py [hunk 3]
```diff
@@ -1,28 +1,59 @@
...
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2012-TODAY David BEAL Akretion <http://www.akretion.com>.
+# © 2012-TODAY Sébastien BEAU Akretion <http://www.akretion.com>.
+# © 2013-2014 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 from openerp import models, fields, api, exceptions, _
+from openerp.exceptions import Warning as UserError
```
Comment (guewen) [inline base_delivery_carrier_label/models/stock.py]: you could directly import UserError: `from openerp.exceptions import UserError`

**Comment by guewen** [comment]
`base_delivery_carrier_label/models/stock.py:6:1: F401 'exceptions' imported but unused`

**Comment by guewen** [comment]
Error in the tests:

``` python

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: ERROR: test_get_weight (openerp.addons.base_delivery_carrier_label.tests.test_get_weight.TestGetWeight)

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: ` Test quant.package.get_weight and pack.operation.get_weight.

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: Traceback (most recent call last):

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: `   File "/home/travis/build/OCA/carrier-delivery/base_delivery_carrier_label/tests/test_get_weight.py", line 86, in test_get_weight

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: `     picking = self._generate_picking(products)

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: `   File "/home/travis/build/OCA/carrier-delivery/base_delivery_carrier_label/tests/test_get_weight.py", line 76, in _generate_picking

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: `     order.action_button_confirm()

2016-04-26 15:32:36,690 23971 ERROR openerp_test openerp.addons.base_delivery_carrier_label.tests.test_get_weight: ` AttributeError: 'sale.order' object has no attribute 'action_button_confirm'

```

**Comment by guewen** [comment]
> AttributeError: 'sale.order' object has no attribute 'action_button_confirm'

It is `action_confirm` now

File: base_delivery_carrier_label/views/stock.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<openerp>
+  <data>
```
Comment (mdietrichc2c) [inline base_delivery_carrier_label/views/stock.xml]: `<odoo>`

File: base_delivery_carrier_label/models/__init__.py (hunk 2 above)
Comment (bealdav) [inline base_delivery_carrier_label/models/__init__.py]: I'm also not sure.

**Comment by yvaucher** [comment]
@alexis-via I don't find from where method `set_pack_weight` is called on model picking.

@bealdav @guewen 
Tests on get_weight are making trouble, a lot has changed to compute weight on 9.0. We now have a weight field already on `stock.quant.package`. Plus logistic units have disapeared, not sure how to replace them, should we add a second weight field on `stock.quant.package` to include the weight of the packaging ?

**Comment by hparfr** [comment]
> Tests on get_weight are making trouble,

It's good: it's an highlight on breaking changes :)

> Plus logistic units have disapeared

Logistic unit weight is needed for the carriers who wants a total weight for the package.
And it's the responsibility of the package to include it his calculation of his own weight.

**Comment by yvaucher** [comment]
@hparfr Thus have you any clue if in official addons of 9.0 we can still manage logistic units or if we need ... again ... to recreate an object they dropped ?

**Comment by yvaucher** [comment]
Ok after more thinking, I will add a new field on `stock.quant.package` object named `total_weight` that allow to override the computed `weight` field value (if not defined value will still be computed as it is now otherwise we put total_weight in it).

Thus in case you weight your final package before delivery you will only have to set the value the balance gave you.

As I understand the weight value is important for the colissimo carrier.

**Comment by hparfr** [comment]
Suggestion of simplifications : 
- remove generate_default_label method
- stop the partial use of packages and picking instances *

\* For instance, change

``` py
 def generate_labels(self, package_ids=None):
```

to 

``` py
def generate_labels(self):
```

> A list of package ids can be given, in that case it will generate  the labels only of these packages.

This feature adds complexity in the juggling between packages_id and i'm not sure about the business needs on this one.

What do you think @yvaucher ?

**Comment by yvaucher** [comment]
@hparfr 

`package_ids` parameter is used in `delivery_carrier_label_dispatch` module.

And `generate_default_label` was a hook used in `delivery_carrier_label_default_webkit` which probably needs to be ported in Qweb.

What can be questionable is to keep both logic of label per picking or per package. What we could do is forcing the user to make packages thus we have only a single case to manage.

**Comment by hparfr** [comment]
> And generate_default_label was a hook used in delivery_carrier_label_default_webkit which probably needs to be ported in Qweb.

`delivery_carrier_label_default_webkit` has not been ported in v8, we may drop support of it for v9 by removing `generate_default_label.` from base module. 

## 

> What we could do is forcing the user to make packages thus we have only a single case to manage.

👍

**Comment by yvaucher** [comment]
Fixed weight calculs and tests. I removed the logic with logistic units providing a weight for the container.

File: README.md [hunk 5]
```diff
@@ -13,13 +13,18 @@ You'll find:
  - Generation of shipping labels for specific carrier (PostLogistics, ...)
 
 [//]: # (addons)
+Available addons
+---------------
+addon | version | summary
+--- | --- | ---
+[base_delivery_carrier_label](base_delivery_carrier_label/) | 9.0.1.2.0 | Base module for carrier labels
```
Comment (lasley) [inline README.md]: 9.0.1.0.0

File: base_delivery_carrier_label/__openerp__.py [hunk 6]
```diff
@@ -1,40 +1,22 @@
...
-          'res_config_view.xml',
-          'security/ir.model.access.csv',
-          ],
+ 'data': [
+     'views/delivery.xml',
+     'views/stock.xml',
+     'views/res_config.xml',
+     'security/ir.model.access.csv',
+ ],
  'tests': [],
```
Comment (lasley) [inline base_delivery_carrier_label/__openerp__.py:18]: Remove empty key

File: base_delivery_carrier_label/models/delivery.py [hunk 7]
```diff
@@ -1,24 +1,7 @@
...
-#    GNU Affero General Public License for more details.
-#
-#    You should have received a copy of the GNU Affero General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
+# © 2012-TODAY Sébastien BEAU Akretion <http://www.akretion.com>.
+# © 2014 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 from openerp import models, fields, api
```
Comment (lasley) [inline base_delivery_carrier_label/models/delivery.py]: Would you mind splitting these models into one per file?

File: base_delivery_carrier_label/models/stock.py [hunk 8]
```diff
@@ -1,40 +1,102 @@
 # -*- coding: utf-8 -*-
```
Comment (lasley) [inline base_delivery_carrier_label/models/stock.py]: Would you mind splitting these models into one per file?

File: README.md (hunk 5 above)
Comment (pedrobaeza) [inline README.md]: This it automatically done by a bot, so there's no need to change it

Comment (lasley) [inline README.md]: TIL - Hooray automation!

File: base_delivery_carrier_label/models/delivery.py (hunk 7 above)
Comment (yvaucher) [inline base_delivery_carrier_label/models/delivery.py]: ~~Can you explain why should I do that?~~
~~I prefer to keep it that way as it is in most OCA code base~~

Ok I thought you wanted me to split the imports.

Those models are completely related. Thus I prefer to keep them in the same file.

File: base_delivery_carrier_label/models/stock.py (hunk 8 above)
Comment (yvaucher) [inline base_delivery_carrier_label/models/stock.py]: IMO this is readable enough this way.

File: base_delivery_carrier_label/models/delivery.py (hunk 7 above)
Comment (lasley) [inline base_delivery_carrier_label/models/delivery.py]: Hi @yvaucher - OCA guidlines say individual files though which is mentioned in the differences section between OCA and Odoo guidelines, as well as a few other places - https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#differences-with-odoo-guidelines

**Comment by yvaucher** [comment]
Splitted python files by model

**Comment by yvaucher** [comment]
I proceed with this PR as on itself it has no use and is only a dependency for others thus unlocking dev for other modules.

**Pull request merged**